Create a driver to update StageMode Interactions based on Gesture Input

Phipson opened the pull request:
#### 0a89daa335939130198597b18646c556f00e2abb
<pre>
Create a driver to update StageMode Interactions based on Gesture Input
<a href="https://bugs.webkit.org/show_bug.cgi?id=287820">https://bugs.webkit.org/show_bug.cgi?id=287820</a>
<a href="https://rdar.apple.com/145006262">rdar://145006262</a>

Reviewed by Richard Robinson, Elliott Williams, and Wenson Hsieh.

This PR adds the StageModeInteractionDriver, which is a helper class object used in the ModelProcessModelPlayerProxy
to control how the model responds to the stagemode transform updates, based on the applied stagemode attribute. When
stagemode is set to &apos;none&apos; we ignore all transforms passed to the player proxy. When stagemode is set to &apos;orbit&apos; we
use the translation from the transform to rotate the model. To separate stagemode updates from Javascript updates,
we introduce an Interaction Container that wraps around the model entity, on which the stagemode updates will be applied.

The driver will be implemented in <a href="https://bugs.webkit.org/show_bug.cgi?id=287601.">https://bugs.webkit.org/show_bug.cgi?id=287601.</a>

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::beginStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::endStageModeInteraction):
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
(WebKit::ModelProcessModelPlayerProxy::applyStageModeOperationToDriver):
- Updates the callbacks to use the interaction driver
- Also adds helper function to update the driver&apos;s internal state based on the stagemode attribute

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift: Added.
(WKStageModeInteractionDriver.stageModeOperation):
(WKStageModeInteractionDriver.interactionDidBegin(_:)):
(WKStageModeInteractionDriver.interactionDidUpdate(_:)):
(WKStageModeInteractionDriver.interactionDidEnd):
(WKStageModeInteractionDriver.operationDidChange(_:)):
- Adds the necessary APIs to apply the transform to the interaction target.

* Source/WebKit/WebKitSwift/StageMode/WKStageModeBridging.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: <a href="https://commits.webkit.org/291258@main">https://commits.webkit.org/291258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d531d57c168a64718af7a5db61e31b07dd24a8ad

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92433 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97417 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42940 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20435 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70829 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/28291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9315 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83706 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51158 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9014 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42271 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1287 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99443 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19484 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19734 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79583 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79107 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19606 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23647 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19468 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19155 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22615 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20895 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.swift:
Normally we'd want this file to live inside the `WebKitSwift` subdirectory since its target membership is with WKS. Since you've already filed https://bugs.webkit.org/show_bug.cgi?id=288728 to track moving its membership back to WebKit proper, I'm OK with leaving it here.

emw-apple commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
ditto

emw-apple commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -173,7 +174,7 @@ class ModelProcessModelPlayerProxy final
 
     // For interactions
-    REPtr<REEntityRef> m_interactionContainerEntity;
+    RetainPtr<WKStageModeInteractionDriver> m_stageModeInteractionDriver;
```
Are you missing a `#import "WKStageMode.h"` here?

Phipson replied:
No we already forward declare this in L52 so this works and compiles!

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -2447,6 +2447,9 @@
 		ED82A7F2128C6FAF004477B3 /* WKBundlePageOverlay.h in Headers */ = {isa = PBXBuildFile; fileRef = 1A22F0FF1289FCD90085E74F /* WKBundlePageOverlay.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		EEA52F482D70545300D578B5 /* WKStageMode.h in Headers */ = {isa = PBXBuildFile; fileRef = EE0D3CDD2D3709BE00072978 /* WKStageMode.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		EEA52F492D7055A700D578B5 /* WKStageMode.swift in Sources */ = {isa = PBXBuildFile; fileRef = EE0D3CDC2D3709BE00072978 /* WKStageMode.swift */; };
+		EEA52F4A2D705A2300D578B5 /* WKStageMode.h in Headers */ = {isa = PBXBuildFile; fileRef = EE0D3CDD2D3709BE00072978 /* WKStageMode.h */; };
```
Building `WKStageMode.h` as part of two targets seems wrong. I think it should only have target membership in WebKit. WebKitSwift can and does access WebKit's headers.

whsieh commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
```diff
@@ -0,0 +1,52 @@
+@interface WKStageModeInteractionDriver : NSObject
+@property (nonatomic, readonly) REEntityRef interactionContainerRef;
+
+- (instancetype)initWithModel:(WKSRKEntity * _Nonnull)model container:(REEntityRef)container NS_SWIFT_NAME(init(with:container:));
```
Nit - no need for _Nonnull here, since it's all in an `NS_ASSUME_NONNULL_` section.

whsieh reviewed (commented):
Seems reasonable, modulo Elliott's comments

Phipson committed:
Create a driver to update StageMode Interactions based on Gesture Input
https://bugs.webkit.org/show_bug.cgi?id=287820
rdar://145006262

Reviewed by Richard Robinson, Elliott Williams, and Wenson Hsieh.

This PR adds the StageModeInteractionDriver, which is a helper class object used in the ModelProcessModelPlayerProxy
to control how the model responds to the stagemode transform updates, based on the applied stagemode attribute. When
stagemode is set to 'none' we ignore all transforms passed to the player proxy. When stagemode is set to 'orbit' we
use the translation from the transform to rotate the model. To separate stagemode updates from Javascript updates,
we introduce an Interaction Container that wraps around the model entity, on which the stagemode updates will be applied.

The driver will be implemented in https://bugs.webkit.org/show_bug.cgi?id=287601.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::beginStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::endStageModeInteraction):
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
(WebKit::ModelProcessModelPlayerProxy::applyStageModeOperationToDriver):
- Updates the callbacks to use the interaction driver
- Also adds helper function to update the driver's internal state based on the stagemode attribute

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift: Added.
(WKStageModeInteractionDriver.stageModeOperation):
(WKStageModeInteractionDriver.interactionDidBegin(_:)):
(WKStageModeInteractionDriver.interactionDidUpdate(_:)):
(WKStageModeInteractionDriver.interactionDidEnd):
(WKStageModeInteractionDriver.operationDidChange(_:)):
- Adds the necessary APIs to apply the transform to the interaction target.

* Source/WebKit/WebKitSwift/StageMode/WKStageModeBridging.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: https://commits.webkit.org/291258@main
